### PR TITLE
[JAVA]  java.net.http.HttpClient (java native) form data generator

### PR DESCRIPTION
**Pull request by Rattenkrieg**
Java native `java.net.http.HttpClient` generator was lacking form data support. This PR introduces handling of form-data parameters.
Besides usual contribution routines I've ran `mvn integration-test -f ./samples/client/petstore/java/native/pom.xml`.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request. @bbdouglas  @sreeshas  @jfiala  @lukoyanov  @cbornet  @jeff9finger  @karismann  @Zomzog  @lwlee2608

### CODE REVIEW

**Comment by wing328** [comment]
@Rattenkrieg thanks for the PR. 

Do you mind pulling the latest master to your branch as I've added a test spec for form data (https://github.com/OpenAPITools/openapi-generator/pull/14587) ?

Then you can add a test in `samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java`

**Comment by Rattenkrieg** [comment]
@wing328 done
However there is an issue with test setup i.e. running `mvn integration-test -f ./samples/client/echo_api/java/native/pom.xml` or any other like `apache-httpclient` fails since there is no echo server listening :3000.

**Comment by wing328** [comment]
can you please add a test in samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java?

there's an endpoint that tests form parameters

**Comment by Rattenkrieg** [comment]
> can you please add a test in samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java?
> 
> there's an endpoint that tests form parameters
> 
> (for test values of the form data, please use `Hello World` to ensure we get `Hello%20World`)

I believe I've placed it in right file. https://github.com/OpenAPITools/openapi-generator/pull/14554/commits/7aadeb3232077172f9c5e402fcbb7a64c9e8ab67#diff-43b8a464f4da55b9453d495bd235c5091b9a0de7a0d4ab0cfbc9fc66386cbe00R260
I will change values to `Hello World`

File: samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java [hunk 1]
```diff
@@ -245,4 +246,25 @@ public void testDefaultValuesSerializationWithIncorrectDefaultValues() throws IO
...
+    /**
+     * Test form parameter(s)
+     *
+     * Test form parameter(s)
+     *
+     * @throws ApiException
+     *          if the Api call fails
+     */
+    @Test
+    public void testFormIntegerBooleanStringTest() throws ApiException {
```
Comment (Rattenkrieg) [inline samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: Here goes the test, however I don't know how to check response for form data parameters. Can't find anything that I could use in `EchoServerResponseParser`.

Comment (wing328) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: Right, I think we need to add a field to store the response body (e.g. `body`) and update the following code accordingly:
```
        for (String line : lines) {
            if (firstLine) {
                String[] items = line.split(" ");
                this.method = items[0];
                this.path = items[1];
                this.protocol = items[2];
                firstLine = false;
                continue;
            }

            // store the header key-value pair in headers
            String[] keyValue = line.split(": ");
            if (keyValue.length == 2) { // skip blank line, non key-value pair
                this.headers.put(keyValue[0], keyValue[1]);
            }
            
            // find the blank line and then store the rest into `body`
        }
```

Is that something you can help with ? 🙏

Comment (Rattenkrieg) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: Sure, let me try.

Comment (Rattenkrieg) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: I finally found (in CI setup) how to run echo server. 
```yaml
        run: |
          git clone https://github.com/wing328/http-echo-server -b openapi-generator-test-server
          (cd http-echo-server && npm install && npm start &)
```
🎉

Comment (wing328) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: > I finally found (in CI setup) how to run echo server.

We also mention it in https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests#echo-server

File: samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java [hunk 2]
```diff
@@ -245,4 +246,25 @@ public void testDefaultValuesSerializationWithIncorrectDefaultValues() throws IO
...
+        Integer integerForm = 1337;
+        Boolean booleanForm = true;
+        String stringForm = "Hello World";
+        String response = formApi.testFormIntegerBooleanString(integerForm, booleanForm, stringForm);
+
+        org.openapitools.client.EchoServerResponseParser p = new org.openapitools.client.EchoServerResponseParser(response);
+        Assert.assertEquals("/form/integer/boolean/string", p.path);
+        Assert.assertTrue(p.body.contains("Hello%20World"));
```
Comment (wing328) [inline samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:268]: one last suggestion: can we do an exact match using assertEquals instead?

We want to catch any potential change in the body

File: samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java (hunk 1 above)
Comment (wing328) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: Sorry. I thought I shared the wiki link earlier but I didn't.

Comment (Rattenkrieg) [reply samples/client/echo_api/java/native/src/test/java/org/openapitools/client/CustomTest.java:260]: @wing328 done

**Comment by ChristianCiach** [comment]
This adds an extra dependency to `org.apache.httpcomponents:httpmime` for everyone, even if you don't need multiparty support. Is this dependency really necessary? Can it at least be opt-in for people who actually need this?

I like the `native` generator for the very reason that it is light on dependencies.

**Comment by ChristianCiach** [comment]
Even worse, this transitively pulls in the full `org.apache.httpcomponents:httpclient`, which also depends on numerous other apache libraries. At this point, what is even the point of the `native` library if it depends on the apache http client?

**Comment by wing328** [comment]
can you please give it another try with the latest master? i've merged a PR to add the spotless plugin that can help remove unused import (command can be found in the build file).

**Comment by ChristianCiach** [comment]
I am sorry, but I don't see how removing unused imports can help here. As long as the `build.gradle` references `org.apache.httpcomponents:httpmime` as a dependency, the `native` generator is mostly useless. If you are okay with having the apache httpclient on the classpath, then you can just as well use the `apache-httpclient` generator.

**Comment by wing328** [comment]
You're right. Looks like the best option at the moment is to add an option to skip the form data related code/import introduced by this PR.

Does it sound good to you?

**Comment by ChristianCiach** [comment]
Yes, that would be fine for me! I would prefer to have this disabled by default, but as long as it will be possible to exclude the apache stuff, I will be okay with this either way :)

**Comment by wing328** [comment]
I've filed https://github.com/OpenAPITools/openapi-generator/pull/14670 with minor enhancements. No need to add another option.

@ChristianCiach please review to see if it meets your requirement.

@Rattenkrieg  thanks for filing this PR, which is a good start.

Have a nice weekend.

**Comment by davidmoten** [comment]
java.net.http.HttpClient requires Java 11, and the lack of support for Java 8 now is not documented anywhere that I can find. Was this really the intent of the project owners? Not even mentioned in release notes that I can find.

**Pull request merged**